### ISSUE TITLE
After upgrading ClickHouse from v20.4.4.18 to v20.8.12.2 server refuses to start. (Corruption)

### ISSUE THREAD

**Issue by artem-pershin** [ClickHouse/ClickHouse#23779]
I have a couple of tables with data in my docker-containered (with volumes as persistent storage) ClickHouse database. Recently i've upgraded from `v20.4.4.18` to `v20.8.12.2` with data remained. Typical table:
```
CREATE TABLE IF NOT EXISTS typical_tbl
(
       id UInt64,
       someString String,
       someFloat  Float64,
       INDEX someStringIndex someString TYPE set(0) GRANULARITY 1,
       INDEX someFloatIndex someFloat TYPE minmax GRANULARITY 1
)
ENGINE = MergeTree()
PARTITION BY (ceil(id / 1000) % 10)
ORDER by (id, someString)
SETTINGS index_granularity = 4096
```

typical error in logs (after a couple of such errors container terminates):
```
2021.04.29 16:05:05.568532 [ 37 ] {} <Error> base.typical_tbl: Part /var/lib/clickhouse/data/base/typical_tbl/3_4_14776_11 is broken. Looking for parts to replace it.
2021.04.29 16:05:05.568554 [ 37 ] {} <Error> base.typical_tbl: Detaching broken part /var/lib/clickhouse/data/base/typical_tbl/3_4_14776_11 because it covers less than 2 parts. You need to resolve this manually
2021.04.29 16:05:05.569975 [ 33 ] {} <Error> DB::MergeTreeData::loadDataParts(bool)::<lambda()>: Code: 33, e.displayText() = DB::Exception: Cannot read all data. Bytes read: 2. Bytes expected: 8., Stack trace (when copying this message, always include the lines below):
0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x18ccf210 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xe62f32d in /usr/bin/clickhouse
2. DB::ReadBuffer::readStrict(char*, unsigned long) @ 0xe67356f in /usr/bin/clickhouse
3. DB::DataTypeNumberBase<double>::deserializeBinary(DB::Field&, DB::ReadBuffer&) const @ 0x1562cd71 in /usr/bin/clickhouse
4. DB::MergeTreePartition::load(DB::MergeTreeData const&, std::__1::shared_ptr<DB::IDisk> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x161f2dce in /usr/bin/clickhouse
5. DB::IMergeTreeDataPart::loadPartitionAndMinMaxIndex() @ 0x160b0310 in /usr/bin/clickhouse
6. DB::IMergeTreeDataPart::loadColumnsChecksumsIndexes(bool, bool) @ 0x160b623a in /usr/bin/clickhouse
7. ? @ 0x160f6aa2 in /usr/bin/clickhouse
8. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xe65d697 in /usr/bin/clickhouse
9. ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()&&...)::'lambda'()::operator()() const @ 0xe65de0a in /usr/bin/clickhouse
10. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xe65cba7 in /usr/bin/clickhouse
11. ? @ 0xe65b093 in /usr/bin/clickhouse
12. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
13. clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
 (version 20.8.12.2 (official build))
```

i've tried to revert ClickHouse version back to `20.4.4.18` and it successfully starts with all data remained. Then i've tried to drop all indices and then upgrade version, but it didn't help.

**Comment by tavplubix** [comment]
Duplicate of #23508
Also related to #23700

**Comment by alexey-milovidov** [comment]
The issue is fixed in #24157.
The fix is backported to versions 21.5 and 21.6.
